src/treeview.ts

```diff
@@ -0,0 +1,186 @@
+      context
+    );
+    allProviders.set(viewId, provider);
+    const view = window.createTreeView(viewId, {
+      treeDataProvider: provider,
+      showCollapseAll: true
+    });
+    allViews.set(viewId, view);
+
+    // Notify the server about view visibility changes
+    const onDidChangeVisibility = view.onDidChangeVisibility(e => {
+      client.sendNotification(MetalsTreeViewVisibilityDidChange.type, {
+        viewId: viewId,
+        visible: e.visible
+      });
+    });
+
+    return [view, onDidChangeVisibility];
+  });
+
+  // Update tree nodes on server notificiations
+  client.onNotification(MetalsTreeViewDidChange.type, params => {
+    params.nodes.forEach(node => {
+      let provider = allProviders.get(node.viewId);
+      if (!provider) return;
+      if (node.nodeUri) {
+        provider.items.set(node.nodeUri, node);
+      }
+      if (node.nodeUri) {
+        provider.didChange.fire(node.nodeUri);
+      } else {
+        provider.didChange.fire(undefined);
+      }
+    });
+  });
+
+  return {
+    disposables: ([] as Disposable[]).concat(...disposables),
+    // Server requested to reveal a tree view node.
+    reveal(params: MetalsRevealTreeView): void {
+      const view = allViews.get(params.viewId);
+      if (view) {
+        view.reveal(params.uri, {
+          expand: params.expand || 3,
+          select: params.select || false,
+          focus: params.focus || true
+        });
+      } else {
+        out.appendLine(`unknown view: ${params.viewId}`);
+      }
+    }
+  };
+}
+
+/**
+ * A tree view data provider with URI-formatted keys.
+ *
+ * The URI-formatted key maps to a `TreeViewNode` value which contains the
+ * metadata about that tree view node such as label, tooltip and icon.
+ *
+ * This data provider is implemented as a proxy by forwarding request about
+ * node children and parents to the Metals server.
+ */
+class MetalsTreeDataProvider implements TreeDataProvider<string> {
+  didChange = new EventEmitter<string>();
+  onDidChangeTreeData?: Event<string> = this.didChange.event;
+  items: Map<string, TreeViewNode> = new Map();
+  constructor(
+    readonly client: LanguageClient,
+    readonly out: OutputChannel,
+    readonly viewId: string,
+    readonly views: Map<string, MetalsTreeDataProvider>,
+    readonly context: ExtensionContext
+  ) {}
+
+  // Populate TreeItem based on cached children response from the server.
+  getTreeItem(uri: string): TreeItem {
+    const item = this.items.get(uri);
+    if (!item) return {};
+    const result: TreeItem = {
+      label: item.label,
+      id: item.nodeUri,
+      resourceUri:
+        item.nodeUri && item.nodeUri.indexOf(".") > 0
+          ? Uri.parse(item.nodeUri)
+          : undefined,
+      collapsibleState:
+        item.collapseState == "expanded"
+          ? TreeItemCollapsibleState.Expanded
+          : item.collapseState == "collapsed"
+          ? TreeItemCollapsibleState.Collapsed
+          : TreeItemCollapsibleState.None,
```

Comment:
can we move this to a utility function?

```ts
function toTreeItemCollapsibleState(s: TreeViewNode["collapseState"]): TreeItemCollapsibleState {
  switch (s) {
    case "expanded": return TreeItemCollapsibleState.Expanded;
    case "collapsed": return TreeItemCollapsibleState.Collapsed;
    case "none": return TreeItemCollapsibleState.None;
  }
}
```

Reply:
Great idea, much clearer. I first tried to implement this helper method on the interface itself but that didn't work (which makes sense when I think of the JS runtime but feels unsatisfying given my JVM background).